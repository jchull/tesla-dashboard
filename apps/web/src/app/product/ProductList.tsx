import React, { useState } from 'react';
import { useDispatch } from 'react-redux';

import { Vehicle } from '@tesla-dashboard/types';

import { ProductListItem } from './ProductListItem';
import { fetchProductListAction, ProductState, selectProduct } from './actions';

import './product.style.scss';
import { NavLink } from 'react-router-dom';

export const ProductList: React.FC<ProductState> = (props: ProductState) => {
  const dispatch = useDispatch();

  const [collapsed, setCollapsed] = useState(false);

  React.useEffect(() => {
    dispatch(fetchProductListAction());
  }, []);

  return (
    <div className={`product-list ${collapsed ? 'collapsed' : ''}`}>
      {props.products.map((product: Vehicle) => (
        <ProductListItem
          product={product}
          key={product._id}
          handleSelection={() => dispatch(selectProduct(product._id))}
          selected={props.selectedProductId === product._id}
        />
      ))}
      <NavLink to="/tesla-account" className="add-more">
        add more linked accounts
      </NavLink>
      <button className="toggle" onClick={() => setCollapsed(!collapsed)} />
    </div>
  );
};
