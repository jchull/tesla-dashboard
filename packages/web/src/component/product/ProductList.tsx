import React, { useState } from 'react'
import { useDispatch } from 'react-redux'

import { Product } from '@teslapp/common'

import { ProductListItem } from './ProductListItem'
import { fetchProductListAction, ProductState, selectProduct } from './actions'

import './product.style.scss'
import { NavLink } from 'react-router-dom'

export const ProductList: React.FC<ProductState> = (props: ProductState) => {
  const dispatch = useDispatch()
  
  const [collapsed, setCollapsed] = useState(false)

  React.useEffect(() => {
    dispatch(fetchProductListAction())
  }, [])

  return (
    <div className={`product-list ${collapsed? 'collapsed' : ''}`}>
      {props.products.map((product: Product) => (
        <ProductListItem
          product={product}
          key={product._id}
          handleSelection={() => dispatch(selectProduct(product._id))}
          selected={props.selectedProductId === product._id}
        />
      ))}
      <NavLink to="/tesla-account" className="add-more">
        add more linked accounts
      </NavLink>
      <button className="toggle" onClick={() => setCollapsed(!collapsed)}></button>
    </div>
  )
}
