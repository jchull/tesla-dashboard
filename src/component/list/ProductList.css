.product-list {
    border-bottom: 1px solid #f4f4f4;
    box-shadow: 1px 2px 3px rgba(111,113,114, 0.3);
}

.product-list .list-item {
    background: #f7f7f7;
    width: 320px;
    padding: 6px;
    border: none;
    margin: 6px;
}
