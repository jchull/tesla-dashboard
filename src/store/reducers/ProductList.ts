import {ACTION_TYPES, ProductListAction} from '../actions';
import {IVehicle} from 'tesla-dashboard-api';
import {ProductListState} from '../types/state';

const initialState = {
  products: [] as IVehicle[]
} as ProductListState;

export function productListReducer(state = initialState, action: ProductListAction): ProductListState {
  switch (action.type) {
    case ACTION_TYPES.UPDATE_PRODUCT_LIST:
      return {
        products: action.productList
      };
      break;

    default:
      return state;
  }
}