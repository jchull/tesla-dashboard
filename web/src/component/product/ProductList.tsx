import React from 'react';
import {useDispatch} from 'react-redux';

import {Product} from '@model/index';

import {ProductListItem} from './ProductListItem';
import {fetchProductListAction, ProductState, selectProduct} from './actions';

import './product.style.scss';


export const ProductList: React.FC<ProductState> = (props: ProductState) => {
  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(fetchProductListAction());
  }, []);

  return (
      <div className="product-list">
        {
          props.products.map(
              (product: Product) => <ProductListItem product={product}
                                                     key={product._id}
                                                     handleSelection={() => dispatch(selectProduct(product._id))}
                                                     selected={props.selectedProductId === product._id}/>)
        }
      </div>
  );
};

