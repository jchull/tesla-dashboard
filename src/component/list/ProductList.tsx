import React from 'react';
import {ProductListItem} from './ProductListItem';
import {queryService} from '@service/Services';
import {VehicleView} from '../view/VehicleView';
import './ProductList.scss';
import {IVehicle} from 'tesla-dashboard-api';
import {AppState} from '../../store/types/state';
import {useDispatch, useSelector} from 'react-redux';
import {ACTION_TYPES} from '../../store/actions';


export const ProductList: React.FC = () => {
  const productListState = useSelector((store: AppState) => store.productList);
  const dispatcher = useDispatch();

  const [selectedProduct, setSelectedProduct] = React.useState({} as IVehicle);

  const productSelectionHandler = (product: IVehicle) => setSelectedProduct(Object.assign({}, product));

  React.useEffect(() => {
    const fetchProductList = async () => {
      console.log('fetching products...');
      const productList = await queryService.getProducts();
      dispatcher({type: ACTION_TYPES.UPDATE_PRODUCT_LIST, productList});
      if (productList && productList.length) {
        setSelectedProduct(productList[0]);
      }
    };

    fetchProductList();
    setInterval(fetchProductList, 60000);
  }, []);

  React.useEffect(() => {
    if (selectedProduct.vin) {
      // TODO: action upon product selection
      // maybe start data refresh for that product only or something
    }
  }, [selectedProduct]);

  return (
      <div>
        <div className="product-list">
          {
            productListState.products && productListState.products.map(
                product => <ProductListItem product={product}
                                            key={product.vin}
                                            handleSelection={productSelectionHandler}
                                            selected={selectedProduct && selectedProduct.vin === product.vin}/>)
          }
        </div>
        <div className="product-view">
          {
            productListState.products && selectedProduct && <VehicleView vehicle={selectedProduct}/>
          }
        </div>
      </div>

  );
};

