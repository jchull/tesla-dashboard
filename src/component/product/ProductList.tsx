import React from 'react';
import {ProductListItem} from './ProductListItem';
import {SessionListView} from '../session/SessionListView';
import './ProductList.scss';
import {IVehicle} from 'tesla-dashboard-api';
import {AppState} from '../../store/types/state';
import {useDispatch, useSelector} from 'react-redux';
import {fetchProductListAction, fetchProductListStart, selectProduct} from './actions';


export const ProductList: React.FC = () => {
  const productListState = useSelector((store: AppState) => store.productList);
  const dispatch = useDispatch();

  const productSelectionHandler = (product: IVehicle) => dispatch(selectProduct(product));

  React.useEffect(() => {
    dispatch(fetchProductListAction());
  }, []);

  return (
      <div>
        <div className="product-list">
          {
            productListState.products && productListState.products.map(
                product => <ProductListItem product={product}
                                            key={product.vin}
                                            handleSelection={productSelectionHandler}
                                            selected={productListState.selectedProduct && productListState.selectedProduct.vin === product.vin}/>)
          }
        </div>
        <div className="product-view">
          <SessionListView/>
        </div>
      </div>

  );
};

