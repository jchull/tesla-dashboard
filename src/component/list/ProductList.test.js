import { act, render } from '@testing-library/react';
import pretty          from 'pretty';
import React           from 'react';


import { ProductList } from './ProductList';

let wrapper;

beforeEach(() => {
});

afterEach(() => {
  if(wrapper) {
    if(wrapper.unmount) {
      wrapper.unmount();
    }
    if(wrapper.container) {
      wrapper.container = null;
    }
  }
});


it('renders empty product list', () => {
  act(() => {
    wrapper = render(<ProductList products={[]}/>);
  });
  expect(pretty(wrapper.container.innerHTML))
    .toMatchSnapshot();
});

