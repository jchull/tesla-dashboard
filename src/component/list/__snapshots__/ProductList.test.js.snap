// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty product list 1`] = `
"<div>
  <div class=\\"product-list\\"></div>
  <div class=\\"product-view\\">
    <div>
      <div>
        <div class=\\"session-list-container\\">
          <div class=\\"session-list\\"></div>
          <div class=\\"selected-view message\\">No product selected</div>
        </div>
        <div class=\\"chart-toolbar\\"><button class=\\"on\\"><i class=\\"material-icons\\">battery_charging_full</i></button><button class=\\"on\\"><i class=\\"material-icons\\">directions_car</i></button></div>
      </div>
    </div>
  </div>
</div>"
`;
